account_budget_oca/models/account_budget.py

```diff
@@ -0,0 +1,216 @@
+    )
+    date_from = fields.Date(string="Start Date", required=True)
+    date_to = fields.Date(string="End Date", required=True)
+    paid_date = fields.Date()
+    planned_amount = fields.Float(required=True, digits=0)
+    practical_amount = fields.Float(compute="_compute_practical_amount", digits=0)
+    theoretical_amount = fields.Float(compute="_compute_theoretical_amount", digits=0)
+    percentage = fields.Float(compute="_compute_percentage", string="Achievement")
+    company_id = fields.Many2one(
+        related="crossovered_budget_id.company_id", store=True, readonly=True,
+    )
+
+    @api.depends(
+        "general_budget_id.account_ids", "date_from", "date_to", "analytic_account_id"
+    )
+    def _compute_practical_amount(self):
+        for line in self:
+            result = 0.0
+            acc_ids = line.general_budget_id.account_ids.ids
+            date_to = line.date_to
+            date_from = line.date_from
+            if line.analytic_account_id.id:
+                self.env.cr.execute(
+                    """
+                    SELECT SUM(amount)
+                    FROM account_analytic_line
+                    WHERE account_id=%s
+                        AND (date between %s
+                        AND %s)
+                        AND general_account_id=ANY(%s)""",
+                    (line.analytic_account_id.id, date_from, date_to, acc_ids,),
+                )
+                result = self.env.cr.fetchone()[0] or 0.0
+            line.practical_amount = result
+
+    @api.depends("paid_date", "date_from", "date_to", "planned_amount")
+    def _compute_theoretical_amount(self):
+        today = fields.Datetime.now()
+        for line in self:
+            # Used for the report
+
+            if line.paid_date:
+                if from_string(line.date_to) <= from_string(line.paid_date):
+                    theo_amt = 0.00
+                else:
+                    theo_amt = line.planned_amount
+            else:
+                line_timedelta = from_string(line.date_to) - from_string(line.date_from)
+                elapsed_timedelta = from_string(today) - from_string(line.date_from)
+
+                if elapsed_timedelta.days < 0:
+                    # If the budget line has not started yet, theoretical
+                    # amount should be zero
+                    theo_amt = 0.00
+                elif line_timedelta.days > 0 and from_string(today) < from_string(
+                    line.date_to
+                ):
+                    # If today is between the budget line date_from and
+                    # date_to
+                    theo_amt = (
+                        elapsed_timedelta.total_seconds()
+                        / line_timedelta.total_seconds()
+                    ) * line.planned_amount
+                else:
+                    theo_amt = line.planned_amount
+
+            line.theoretical_amount = theo_amt
```

Comment:
```suggestion
            else:
                theo_amt = 0.00
            line.theoretical_amount = theo_amt
```
part (2/2)

Reply:
sorry @RachidAlassir, i left this totally unattended .. and forgotten :(
thanks @gaikaz for the suggestions ! they are already applied !